### DISCUSSION TITLE
SPDM Cryptography Selection

### DISCUSSION THREAD

**Discussion by steven-bellock** [DMTF/libspdm#121] [Ideas]
This blog post details how the golang TLS library presents cryptography suites to developers. Seems applicable to SPDM as it leverages a lot from TLS.

https://go.dev/blog/tls-cipher-suites

**Comment by jyao1** [comment]
Current LIBSPDM hardcodes the selection in https://github.com/DMTF/libspdm/blob/main/library/spdm_responder_lib/algorithms.c, search *_priority_table.

**Comment by steven-bellock** [comment]
The [OCP attestation specification](https://docs.google.com/document/d/1fzBnA6N9vGYj--Mxit9Whp0TPPyGqgOyBlHdgCN9F_Q/edit#) also has guidelines for cryptography algorithm support.

![image](https://user-images.githubusercontent.com/8474235/136462839-10937f10-a257-483a-afe4-713ade92aae2.png)

**Comment by jyao1** [comment]
PCISIG "CMA ECN" mentioned below requirement:

Responders must, for BaseAsymAlgo, support one or more of the following:
o TPM_ALG_RSASSA_3072
o TPM_ALG_ECDSA_ECC_NIST_P256
o TPM_ALG_ECDSA_ECC_NIST_P384

Requesters and responders must, for MeasurementHashAlgo, support one or both of the following:
o TPM_ALG_SHA_256
o TPM_ALG_SHA_384

**Comment by bhenning10** [comment]
I wrote this chunk of the OCP spec. The idea was to align with both CMA allowed values, so the minimums are based on CMA, and also to align with CNSA, which OCP references. 
Now that I look at this, the Minimum may drop SHA-256 in the next revision as that is not on the CNSA list.

**Comment by jyao1** [comment]
An open question: Do we want to provide a hook to let spdm consumer decide what the priority list is?

E.g. between RSA3072 and ECDSA256, which is preferred?

**Reply by steven-bellock** [reply]
Yes, I think it's fine for libspdm to default to
- ECC is better than RSA.
- Larger keys are better than smaller keys.

but allow the integrator to override that with their own priority algorithm.
